Improved fill hole suggestions

[A lot of work](https://downloads.haskell.org/ghc/latest/docs/html/users_guide/glasgow_exts.html#typed-holes) has gone into making GHC able to produce suggestions and the current code action falls short of making the most of it. The possibilities include:

- Should it use `-fshow-hole-constraints` to make added constraints explicit?
- Should it use `-fshow-docs-of-hole-fits` to show the docs of the suggestion?
- Should it use `-frefinement-level-hole-fits=1` to show almost-valid fits?
- Should it use `-fsort-by-subsumption-hole-fits` ?

noob question:

> Should it use -fshow-hole-constraints to make added constraints explicit?

Do you mean that if there is a suggestion of, for example, `mempty`, in the list of quickfix, we should show this value is from Monoid typeclass?

I'm confused because constraints shown don't seem bound on specific suggestions

For example:

```
Main.hs:8:5: error:
    • Found hole: _ :: a
      Where: ‘a’ is a rigid type variable bound by
               the type signature for:
                 f :: forall a. Monoid a => a
               at Main.hs:7:1-18
    • In the expression: _
      In an equation for ‘f’: f = _
    • Relevant bindings include f :: a (bound at Main.hs:8:1)
      Constraints include Monoid a (from Main.hs:7:1-18) <------------- this line
      Valid hole fits include
        f :: a (bound at Main.hs:8:1)
        mempty :: forall a. Monoid a => a
          with mempty @a
          (imported from ‘Prelude’ at Main.hs:3:1-14
           (and originally defined in ‘GHC.Base’))
```

--- a/ghcide.cabal
+++ b/ghcide.cabal
@@ -101,6 +101,7 @@ library
         BangPatterns
         DeriveFunctor
         DeriveGeneric
+        FlexibleContexts
         GeneralizedNewtypeDeriving
         LambdaCase
         NamedFieldPuns
@@ -223,6 +224,7 @@ benchmark benchHist
         BangPatterns
         DeriveFunctor
         DeriveGeneric
+        FlexibleContexts
         GeneralizedNewtypeDeriving
         LambdaCase
         NamedFieldPuns
@@ -290,6 +292,7 @@ executable ghcide
         BangPatterns
         DeriveFunctor
         DeriveGeneric
+        FlexibleContexts
         GeneralizedNewtypeDeriving
         LambdaCase
         NamedFieldPuns
@@ -360,6 +363,7 @@ test-suite ghcide-tests
         BangPatterns
         DeriveFunctor
         DeriveGeneric
+        FlexibleContexts
         GeneralizedNewtypeDeriving
         LambdaCase
         NamedFieldPuns
@@ -398,6 +402,7 @@ executable ghcide-bench
         BangPatterns
         DeriveFunctor
         DeriveGeneric
+        FlexibleContexts
         GeneralizedNewtypeDeriving
         LambdaCase
         NamedFieldPuns

--- a/session-loader/Development/IDE/Session.hs
+++ b/session-loader/Development/IDE/Session.hs
@@ -35,6 +35,7 @@ import Data.Version
 import Development.IDE.Core.OfInterest
 import Development.IDE.Core.Shake
 import Development.IDE.Core.RuleTypes
+import Development.IDE.GHC.Compat
 import Development.IDE.GHC.Util
 import Development.IDE.Session.VersionCheck
 import Development.IDE.Types.Diagnostics
@@ -56,7 +57,6 @@ import System.FilePath
 import System.Info
 import System.IO
 
-import GHC
 import GHCi
 import DynFlags
 import HscTypes
@@ -181,7 +181,7 @@ loadSession dir = do
                 -> IO ([NormalizedFilePath],(IdeResult HscEnvEq,[FilePath]))
         session args@(hieYaml, _cfp, _opts, _libDir) = do
           (hscEnv, new, old_deps) <- packageSetup args
-          
+
           -- Whenever we spin up a session on Linux, dynamically load libm.so.6
           -- in. We need this in case the binary is statically linked, in which
           -- case the interactive session will fail when trying to load
@@ -196,7 +196,7 @@ loadSession dir = do
             case res of
               Nothing -> pure ()
               Just err -> hPutStrLn stderr $
-                "Error dynamically loading libm.so.6:\n" <> err 
+                "Error dynamically loading libm.so.6:\n" <> err
 
           -- Make a map from unit-id to DynFlags, this is used when trying to
           -- resolve imports. (especially PackageImports)
@@ -575,6 +575,7 @@ setOptions (ComponentOptions theOpts compRoot _) dflags = do
           setIgnoreInterfacePragmas $
           setLinkerOptions $
           disableOptimisation $
+          setUpTypedHoles $
           makeDynFlagsAbsolute compRoot dflags'
     -- initPackages parses the -package flags and
     -- sets up the visibility for each component.

--- a/session-loader/Development/IDE/Session.hs
+++ b/session-loader/Development/IDE/Session.hs
@@ -105,7 +105,10 @@ loadSession dir = do
     ShakeExtras{logger, eventer, restartShakeSession, withIndefiniteProgress
                ,ideNc, knownFilesVar, session=ideSession} <- getShakeExtras
 
-    IdeOptions{optTesting = IdeTesting optTesting, optCheckProject = CheckProject checkProject } <- getIdeOptions
+    IdeOptions{ optTesting = IdeTesting optTesting
+              , optCheckProject = CheckProject checkProject
+              , optCustomDynFlags
+              } <- getIdeOptions
 
     -- Create a new HscEnv from a hieYaml root and a set of options
     -- If the hieYaml file already has an HscEnv, the new component is
@@ -117,7 +120,7 @@ loadSession dir = do
           -- Parse DynFlags for the newly discovered component
           hscEnv <- emptyHscEnv ideNc libDir
           (df, targets) <- evalGhcEnv hscEnv $
-              setOptions opts (hsc_dflags hscEnv)
+              first optCustomDynFlags <$> setOptions opts (hsc_dflags hscEnv)
           let deps = componentDependencies opts ++ maybeToList hieYaml
           dep_info <- getDependencyInfo deps
           -- Now lookup to see whether we are combining with an existing HscEnv

--- a/src/Development/IDE/GHC/Compat.hs
+++ b/src/Development/IDE/GHC/Compat.hs
@@ -330,7 +330,9 @@ setUpTypedHoles ::DynFlags -> DynFlags
 #if MIN_GHC_API_VERSION(8,6,0)
 setUpTypedHoles df
   = flip gopt_unset Opt_AbstractRefHoleFits    -- too spammy
+#if MIN_GHC_API_VERSION(8,8,0)
   $ flip gopt_unset Opt_ShowDocsOfHoleFits     -- not used
+#endif
   $ flip gopt_unset Opt_ShowMatchesOfHoleFits  -- nice but broken (forgets module qualifiers)
   $ flip gopt_unset Opt_ShowProvOfHoleFits     -- not used
   $ flip gopt_unset Opt_ShowTypeAppOfHoleFits  -- not used

--- a/src/Development/IDE/GHC/Compat.hs
+++ b/src/Development/IDE/GHC/Compat.hs
@@ -327,12 +327,27 @@ dontWriteHieFiles d =
 #endif
 
 setUpTypedHoles ::DynFlags -> DynFlags
-setUpTypedHoles df
-  = df
 #if MIN_GHC_API_VERSION(8,6,0)
-  { refLevelHoleFits = Just 2  -- GHC diagnostics become too slow for level > 2
+setUpTypedHoles df
+  = flip gopt_unset Opt_AbstractRefHoleFits    -- too spammy
+  $ flip gopt_unset Opt_ShowDocsOfHoleFits     -- not used
+  $ flip gopt_unset Opt_ShowMatchesOfHoleFits  -- nice but broken (forgets module qualifiers)
+  $ flip gopt_unset Opt_ShowProvOfHoleFits     -- not used
+  $ flip gopt_unset Opt_ShowTypeAppOfHoleFits  -- not used
+  $ flip gopt_unset Opt_ShowTypeAppVarsOfHoleFits -- not used
+  $ flip gopt_unset Opt_ShowTypeOfHoleFits     -- massively simplifies parsing
+  $ flip gopt_set   Opt_SortBySubsumHoleFits   -- very nice and fast enough in most cases
+  $ flip gopt_unset Opt_SortValidHoleFits
+  $ flip gopt_unset Opt_UnclutterValidHoleFits
+  $ df
+  { refLevelHoleFits = Just 1   -- becomes slow at higher levels
+  , maxRefHoleFits   = Just 10  -- quantity does not impact speed
+  , maxValidHoleFits = Nothing  -- quantity does not impact speed
   }
+#else
+setUpTypedHoles = id
 #endif
+
 
 nameListFromAvails :: [AvailInfo] -> [(SrcSpan, Name)]
 nameListFromAvails as =

--- a/src/Development/IDE/GHC/Compat.hs
+++ b/src/Development/IDE/GHC/Compat.hs
@@ -31,6 +31,7 @@ module Development.IDE.GHC.Compat(
     addIncludePathsQuote,
     getModuleHash,
     getPackageName,
+    setUpTypedHoles,
     pattern DerivD,
     pattern ForD,
     pattern InstD,
@@ -323,6 +324,14 @@ dontWriteHieFiles d =
     gopt_unset d Opt_WriteHie
 #else
     d
+#endif
+
+setUpTypedHoles ::DynFlags -> DynFlags
+setUpTypedHoles df
+  = df
+#if MIN_GHC_API_VERSION(8,6,0)
+  { refLevelHoleFits = Just 2  -- GHC diagnostics become too slow for level > 2
+  }
 #endif
 
 nameListFromAvails :: [AvailInfo] -> [(SrcSpan, Name)]

--- a/src/Development/IDE/Plugin/CodeAction.hs
+++ b/src/Development/IDE/Plugin/CodeAction.hs
@@ -48,8 +48,7 @@ import qualified Data.Text as T
 import Data.Tuple.Extra ((&&&))
 import HscTypes
 import Parser
-import Text.Regex.TDFA (CompOption(..), RegexContext, makeRegexOpts, blankCompOpt, blankExecOpt, matchM, (=~), (=~~))
-import Text.Regex.TDFA.Text(Regex)
+import Text.Regex.TDFA (mrAfter, (=~), (=~~))
 import Outputable (ppr, showSDocUnsafe)
 import DynFlags (xFlags, FlagSpec(..))
 import GHC.LanguageExtensions.Type (Extension)
@@ -143,14 +142,14 @@ suggestAction
   -> Diagnostic
   -> [(T.Text, [TextEdit])]
 suggestAction dflags packageExports ideOptions parsedModule text diag = concat
-    [ suggestAddExtension diag
+   -- Order these suggestions by priority
+    [ suggestAddExtension diag             -- Highest priority
+    , suggestSignature True diag
     , suggestExtendImport dflags text diag
-    , suggestFillHole diag
     , suggestFillTypeWildcard diag
     , suggestFixConstructorImport text diag
     , suggestModuleTypo diag
     , suggestReplaceIdentifier text diag
-    , suggestSignature True diag
     , suggestConstraint text diag
     , removeRedundantConstraints text diag
     , suggestAddTypeAnnotationToSatisfyContraints text diag
@@ -160,7 +159,9 @@ suggestAction dflags packageExports ideOptions parsedModule text diag = concat
     ++ suggestNewImport packageExports pm diag
     ++ suggestDeleteUnusedBinding pm text diag
     ++ suggestExportUnusedTopBinding text pm diag
-    | Just pm <- [parsedModule]]
+    | Just pm <- [parsedModule]
+    ] ++
+    suggestFillHole diag                   -- Lowest priority
 
 
 suggestRemoveRedundantImport :: ParsedModule -> Maybe T.Text -> Diagnostic -> [(T.Text, [TextEdit])]
@@ -205,7 +206,7 @@ suggestDeleteUnusedBinding
         name
         (L (RealSrcSpan l) (ValD (extractNameAndMatchesFromFunBind -> Just (lname, matches)))) =
         case lname of
-          (L nLoc _name) | isTheBinding nLoc -> 
+          (L nLoc _name) | isTheBinding nLoc ->
             let findSig (L (RealSrcSpan l) (SigD sig)) = findRelatedSigSpan indexedContent name l sig
                 findSig _ = []
             in
@@ -234,17 +235,17 @@ suggestDeleteUnusedBinding
 
       -- Second of the tuple means there is only one match
       findRelatedSigSpan1 :: String -> Sig GhcPs -> Maybe (SrcSpan, Bool)
-      findRelatedSigSpan1 name (TypeSig lnames _) = 
+      findRelatedSigSpan1 name (TypeSig lnames _) =
         let maybeIdx = findIndex (\(L _ id) -> isSameName id name) lnames
         in case maybeIdx of
             Nothing -> Nothing
             Just _ | length lnames == 1 -> Just (getLoc $ head lnames, True)
-            Just idx -> 
+            Just idx ->
               let targetLname = getLoc $ lnames !! idx
                   startLoc = srcSpanStart targetLname
                   endLoc = srcSpanEnd targetLname
-                  startLoc' = if idx == 0 
-                              then startLoc 
+                  startLoc' = if idx == 0
+                              then startLoc
                               else srcSpanEnd . getLoc $ lnames !! (idx - 1)
                   endLoc' = if idx == 0 && idx < length lnames - 1
                             then srcSpanStart . getLoc $ lnames !! (idx + 1)
@@ -263,7 +264,7 @@ suggestDeleteUnusedBinding
         name
         (L _ Match{m_grhss=GRHSs{grhssLocalBinds}}) = do
         case grhssLocalBinds of
-          (L _ (HsValBinds (ValBinds bag lsigs))) -> 
+          (L _ (HsValBinds (ValBinds bag lsigs))) ->
             if isEmptyBag bag
             then []
             else concatMap (findRelatedSpanForHsBind indexedContent name lsigs) bag
@@ -278,10 +279,10 @@ suggestDeleteUnusedBinding
         -> [LSig GhcPs]
         -> LHsBind GhcPs
         -> [Range]
-      findRelatedSpanForHsBind 
+      findRelatedSpanForHsBind
         indexedContent
-        name 
-        lsigs 
+        name
+        lsigs
         (L (RealSrcSpan l) (extractNameAndMatchesFromFunBind -> Just (lname, matches))) =
         if isTheBinding (getLoc lname)
         then
@@ -517,51 +518,77 @@ suggestModuleTypo Diagnostic{_range=_range,..}
 
 suggestFillHole :: Diagnostic -> [(T.Text, [TextEdit])]
 suggestFillHole Diagnostic{_range=_range,..}
-{-
-  Relevant bindings include
-    ...
-    extractFitNames :: [Text] -> [Text]
-      (bound at /Users/pepeiborra/scratch/ghcide/src/Development/IDE/Plugin/CodeAction.hs:454:7)
-    (Some bindings suppressed; use -fmax-relevant-binds=N or -fno-max-relevant-binds)
-  Valid hole fits include
-    ...
-    tail :: forall a. [a] -> [a]
-      with tail @Text
-      (imported from ‘Data.List.Extra’ at /Users/pepeiborra/scratch/ghcide/src/Development/IDE/Plugin/CodeAction.hs:45:1-22
-       (and originally defined in ‘GHC.List’))
-    (Some hole fits suppressed; use -fmax-valid-hole-fits=N or -fno-max-valid-hole-fits)
-  Valid refinement hole fits include
-    (++) (_ :: [Text])
-      where (++) :: forall a. [a] -> [a] -> [a]
-      with (++) @Text
-      (imported from ‘Data.List.Extra’ at /Users/pepeiborra/scratch/ghcide/src/Development/IDE/Plugin/CodeAction.hs:45:1-22
-       (and originally defined in ‘GHC.Base’))
-    ...
-    (Some refinement hole fits suppressed; use -fmax-refinement-hole-fits=N or -fno-max-refinement-hole-fits)
--}
-    | Just [_, holeFits, refinementFits]
-    <- matchRegexSingleLine
-        _message
-        ".*Valid (hole fits|substitutions) include(.*)Valid refinement hole fits include(.*)"
-    , Just holeName <- extractHoleName _message
-    = map (proposeHoleFit holeName)
-    $ fromMaybe [] (extractHoleFits holeFits)
-    ++ fromMaybe [] (extractRefFits refinementFits)
-    | Just [_, holeFits]
-    <- matchRegexSingleLine
-        _message
-        ".*Valid (hole fits|substitutions) include(.*)"
-    , Just fits <- extractHoleFits holeFits
-    , Just holeName <- extractHoleName _message
-    = map (proposeHoleFit holeName) fits
-
+    | Just holeName <- extractHoleName _message
+    , (holeFits, refFits) <- processHoleSuggestions (T.lines _message)
+    = map (proposeHoleFit holeName False) holeFits
+    ++ map (proposeHoleFit holeName True) refFits
     | otherwise = []
     where
-      extractHoleName = fmap head . flip matchRegex "Found hole: ([^ ]*)"
-      extractHoleFits = fmap (map (!! 1)) . flip matchRegexAll "^    ([^ ]*) :: (.*)$"
-      extractRefFits x = fmap (map (!! 1)) $ matchRegexAll x "^    ([^ ]+( \\(_ :: .+\\))*)$"
-      proposeHoleFit holeName name =
-          ("replace `" <> holeName <> "` with " <> name, [TextEdit _range name])
+      extractHoleName = fmap head . flip matchRegexUnifySpaces "Found hole: ([^ ]*)"
+      proposeHoleFit holeName parenthise name =
+          ( "replace " <> holeName <> " with " <> name
+          , [TextEdit _range $ if parenthise then parens name else name])
+      parens x = "(" <> x <> ")"
+
+processHoleSuggestions :: [T.Text] -> ([T.Text], [T.Text])
+processHoleSuggestions mm = (holeSuggestions, refSuggestions)
+{-
+    • Found hole: _ :: LSP.Handlers
+
+      Valid hole fits include def
+      Valid refinement hole fits include
+        fromMaybe (_ :: LSP.Handlers) (_ :: Maybe LSP.Handlers)
+        fromJust (_ :: Maybe LSP.Handlers)
+        haskell-lsp-types-0.22.0.0:Language.Haskell.LSP.Types.Window.$sel:_value:ProgressParams (_ :: ProgressParams
+                                                                                                        LSP.Handlers)
+        T.foldl (_ :: LSP.Handlers -> Char -> LSP.Handlers)
+                (_ :: LSP.Handlers)
+                (_ :: T.Text)
+        T.foldl' (_ :: LSP.Handlers -> Char -> LSP.Handlers)
+                 (_ :: LSP.Handlers)
+                 (_ :: T.Text)
+-}
+  where
+    t = id @T.Text
+    holeSuggestions = do
+      -- get the text indented under Valid hole fits
+      validHolesSection <-
+        getIndentedGroupsBy (=~ t " *Valid (hole fits|substitutions) include") mm
+      -- the Valid hole fits line can contain a hole fit
+      holeFitLine <-
+        mapHead
+            (mrAfter . (=~ t " *Valid (hole fits|substitutions) include"))
+            validHolesSection
+      let holeFit = T.strip holeFitLine
+      guard (not $ T.null holeFit)
+      return holeFit
+    refSuggestions = do -- @[]
+      -- get the text indented under Valid refinement hole fits
+      refinementSection <-
+        getIndentedGroupsBy (=~ t " *Valid refinement hole fits include") mm
+      -- get the text for each hole fit
+      holeFitLines <- getIndentedGroups (tail refinementSection)
+      let holeFit = T.strip $ T.unwords holeFitLines
+      guard $ not $ holeFit =~ t "Some refinement hole fits suppressed"
+      return holeFit
+
+    mapHead f (a:aa) = f a : aa
+    mapHead _ [] = []
+
+-- > getIndentedGroups [" H1", "  l1", "  l2", " H2", "  l3"] = [[" H1,", "  l1", "  l2"], [" H2", "  l3"]]
+getIndentedGroups :: [T.Text] -> [[T.Text]]
+getIndentedGroups [] = []
+getIndentedGroups ll@(l:_) = getIndentedGroupsBy ((== indentation l) . indentation) ll
+-- |
+-- > getIndentedGroupsBy (" H" `isPrefixOf`) [" H1", "  l1", "  l2", " H2", "  l3"] = [[" H1", "  l1", "  l2"], [" H2", "  l3"]]
+getIndentedGroupsBy :: (T.Text -> Bool) -> [T.Text] -> [[T.Text]]
+getIndentedGroupsBy pred inp = case dropWhile (not.pred) inp of
+    (l:ll) -> case span (\l' -> indentation l < indentation l') ll of
+        (indented, rest) -> (l:indented) : getIndentedGroupsBy pred rest
+    _ -> []
+
+indentation :: T.Text -> Int
+indentation = T.length . T.takeWhile isSpace
 
 suggestExtendImport :: Maybe DynFlags -> Maybe T.Text -> Diagnostic -> [(T.Text, [TextEdit])]
 suggestExtendImport (Just dflags) contents Diagnostic{_range=_range,..}
@@ -1027,22 +1054,6 @@ matchRegex :: T.Text -> T.Text -> Maybe [T.Text]
 matchRegex message regex = case message =~~ regex of
     Just (_ :: T.Text, _ :: T.Text, _ :: T.Text, bindings) -> Just bindings
     Nothing -> Nothing
-
--- | Returns all the matchgroups
-matchRegexAll :: T.Text -> T.Text -> Maybe [[T.Text]]
-matchRegexAll message regex = message =~~ regex
-
-matchSingleLine
-    :: (RegexContext Regex T.Text target) => T.Text -> T.Text -> Maybe target
-matchSingleLine regex =
-    matchM (makeRegexOpts blankCompOpt{multiline = False} blankExecOpt regex :: Regex)
-
--- | Like 'matchRegex' but treat the input as a single line
---   (dot matches newline, anchors do not treat newline specially)
-matchRegexSingleLine :: T.Text -> T.Text -> Maybe [T.Text]
-matchRegexSingleLine msg regex = do
-  (_ :: T.Text, _ :: T.Text, _ :: T.Text, bindings) <- matchSingleLine regex msg
-  return bindings
 
 setHandlersCodeLens :: PartialHandlers c
 setHandlersCodeLens = PartialHandlers $ \WithMessage{..} x -> return x{

--- a/src/Development/IDE/Plugin/CodeAction.hs
+++ b/src/Development/IDE/Plugin/CodeAction.hs
@@ -48,8 +48,8 @@ import qualified Data.Text as T
 import Data.Tuple.Extra ((&&&))
 import HscTypes
 import Parser
-import Text.Regex.TDFA ((=~), (=~~))
-import Text.Regex.TDFA.Text()
+import Text.Regex.TDFA (CompOption(..), RegexContext, makeRegexOpts, blankCompOpt, blankExecOpt, matchM, (=~), (=~~))
+import Text.Regex.TDFA.Text(Regex)
 import Outputable (ppr, showSDocUnsafe)
 import DynFlags (xFlags, FlagSpec(..))
 import GHC.LanguageExtensions.Type (Extension)
@@ -166,7 +166,7 @@ suggestAction dflags packageExports ideOptions parsedModule text diag = concat
 suggestRemoveRedundantImport :: ParsedModule -> Maybe T.Text -> Diagnostic -> [(T.Text, [TextEdit])]
 suggestRemoveRedundantImport ParsedModule{pm_parsed_source = L _  HsModule{hsmodImports}} contents Diagnostic{_range=_range,..}
 --     The qualified import of ‘many’ from module ‘Control.Applicative’ is redundant
-    | Just [_, bindings] <- matchRegex _message "The( qualified)? import of ‘([^’]*)’ from module [^ ]* is redundant"
+    | Just [_, bindings] <- matchRegexUnifySpaces _message "The( qualified)? import of ‘([^’]*)’ from module [^ ]* is redundant"
     , Just (L _ impDecl) <- find (\(L l _) -> srcSpanToRange l == Just _range ) hsmodImports
     , Just c <- contents
     , ranges <- map (rangesForBinding impDecl . T.unpack) (T.splitOn ", " bindings)
@@ -184,11 +184,11 @@ suggestRemoveRedundantImport ParsedModule{pm_parsed_source = L _  HsModule{hsmod
 
 suggestDeleteUnusedBinding :: ParsedModule -> Maybe T.Text -> Diagnostic -> [(T.Text, [TextEdit])]
 suggestDeleteUnusedBinding
-  ParsedModule{pm_parsed_source = L _ HsModule{hsmodDecls}} 
+  ParsedModule{pm_parsed_source = L _ HsModule{hsmodDecls}}
   contents
   Diagnostic{_range=_range,..}
 -- Foo.hs:4:1: warning: [-Wunused-binds] Defined but not used: ‘f’
-    | Just [name] <- matchRegex _message ".*Defined but not used: ‘([^ ]+)’"
+    | Just [name] <- matchRegexUnifySpaces _message ".*Defined but not used: ‘([^ ]+)’"
     , Just indexedContent <- indexedByPosition . T.unpack <$> contents
       = let edits = flip TextEdit "" <$> relatedRanges indexedContent (T.unpack name)
         in ([("Delete ‘" <> name <> "’", edits) | not (null edits)])
@@ -214,11 +214,11 @@ suggestDeleteUnusedBinding
           _ -> concatMap (findRelatedSpanForMatch indexedContent name) matches
       findRelatedSpans _ _ _ = []
 
-      extractNameAndMatchesFromFunBind 
-        :: HsBind GhcPs 
+      extractNameAndMatchesFromFunBind
+        :: HsBind GhcPs
         -> Maybe (Located (IdP GhcPs), [LMatch GhcPs (LHsExpr GhcPs)])
-      extractNameAndMatchesFromFunBind 
-        FunBind 
+      extractNameAndMatchesFromFunBind
+        FunBind
           { fun_id=lname
           , fun_matches=MG {mg_alts=L _ matches}
           } = Just (lname, matches)
@@ -306,9 +306,9 @@ suggestExportUnusedTopBinding srcOpt ParsedModule{pm_parsed_source = L _ HsModul
 -- Foo.hs:5:1: warning: [-Wunused-top-binds] Defined but not used: type constructor or class ‘F’
 -- Foo.hs:6:1: warning: [-Wunused-top-binds] Defined but not used: data constructor ‘Bar’
   | Just source <- srcOpt
-  , Just [name] <- matchRegex _message ".*Defined but not used: ‘([^ ]+)’"
-                   <|> matchRegex _message ".*Defined but not used: type constructor or class ‘([^ ]+)’"
-                   <|> matchRegex _message ".*Defined but not used: data constructor ‘([^ ]+)’"
+  , Just [name] <- matchRegexUnifySpaces _message ".*Defined but not used: ‘([^ ]+)’"
+                   <|> matchRegexUnifySpaces _message ".*Defined but not used: type constructor or class ‘([^ ]+)’"
+                   <|> matchRegexUnifySpaces _message ".*Defined but not used: data constructor ‘([^ ]+)’"
   , Just (exportType, _) <- find (matchWithDiagnostic _range . snd)
                             . mapMaybe
                                 (\(L l b) -> if maybe False isTopLevel $ srcSpanToRange l
@@ -386,11 +386,11 @@ suggestAddTypeAnnotationToSatisfyContraints sourceOpt Diagnostic{_range=_range,.
 --       In the expression: seq "test" seq "test" (traceShow "test")
 --       In an equation for ‘f’:
 --          f = seq "test" seq "test" (traceShow "test")
-    | Just [ty, lit] <- matchRegex _message (pat False False True)
-                        <|> matchRegex _message (pat False False False)
+    | Just [ty, lit] <- matchRegexUnifySpaces _message (pat False False True)
+                        <|> matchRegexUnifySpaces _message (pat False False False)
             = codeEdit ty lit (makeAnnotatedLit ty lit)
     | Just source <- sourceOpt
-    , Just [ty, lit] <- matchRegex _message (pat True True False)
+    , Just [ty, lit] <- matchRegexUnifySpaces _message (pat True True False)
             = let lit' = makeAnnotatedLit ty lit;
                   tir = textInRange _range source
               in codeEdit ty lit (T.replace lit lit' tir)
@@ -432,9 +432,9 @@ suggestNewDefinition :: IdeOptions -> ParsedModule -> Maybe T.Text -> Diagnostic
 suggestNewDefinition ideOptions parsedModule contents Diagnostic{_message, _range}
 --     * Variable not in scope:
 --         suggestAcion :: Maybe T.Text -> Range -> Range
-    | Just [name, typ] <- matchRegex message "Variable not in scope: ([^ ]+) :: ([^*•]+)"
+    | Just [name, typ] <- matchRegexUnifySpaces message "Variable not in scope: ([^ ]+) :: ([^*•]+)"
     = newDefinitionAction ideOptions parsedModule _range name typ
-    | Just [name, typ] <- matchRegex message "Found hole: _([^ ]+) :: ([^*•]+) Or perhaps"
+    | Just [name, typ] <- matchRegexUnifySpaces message "Found hole: _([^ ]+) :: ([^*•]+) Or perhaps"
     , [(label, newDefinitionEdits)] <- newDefinitionAction ideOptions parsedModule _range name typ
     = [(label, mkRenameEdit contents _range name : newDefinitionEdits)]
     | otherwise = []
@@ -517,44 +517,56 @@ suggestModuleTypo Diagnostic{_range=_range,..}
 
 suggestFillHole :: Diagnostic -> [(T.Text, [TextEdit])]
 suggestFillHole Diagnostic{_range=_range,..}
---  ...Development/IDE/LSP/CodeAction.hs:103:9: warning:
---   * Found hole: _ :: Int -> String
---   * In the expression: _
---     In the expression: _ a
---     In an equation for ‘foo’: foo a = _ a
---   * Relevant bindings include
---       a :: Int
---         (bound at ...Development/IDE/LSP/CodeAction.hs:103:5)
---       foo :: Int -> String
---         (bound at ...Development/IDE/LSP/CodeAction.hs:103:1)
---     Valid hole fits include
---       foo :: Int -> String
---         (bound at ...Development/IDE/LSP/CodeAction.hs:103:1)
---       show :: forall a. Show a => a -> String
---         with show @Int
---         (imported from ‘Prelude’ at ...Development/IDE/LSP/CodeAction.hs:7:8-37
---          (and originally defined in ‘GHC.Show’))
---       mempty :: forall a. Monoid a => a
---         with mempty @(Int -> String)
---         (imported from ‘Prelude’ at ...Development/IDE/LSP/CodeAction.hs:7:8-37
---          (and originally defined in ‘GHC.Base’)) (lsp-ui)
-
-    | topOfHoleFitsMarker `T.isInfixOf` _message = let
-      findSuggestedHoleFits :: T.Text -> [T.Text]
-      findSuggestedHoleFits = extractFitNames . selectLinesWithFits . dropPreceding . T.lines
-      proposeHoleFit name = ("replace hole `" <> holeName <>  "` with " <> name, [TextEdit _range name])
-      holeName = T.strip $ last $ T.splitOn ":" $ head . T.splitOn "::" $ head $ filter ("Found hole" `T.isInfixOf`) $ T.lines _message
-      dropPreceding       = dropWhile (not . (topOfHoleFitsMarker `T.isInfixOf`))
-      selectLinesWithFits = filter ("::" `T.isInfixOf`)
-      extractFitNames     = map (T.strip . head . T.splitOn " :: ")
-      in map proposeHoleFit $ nubOrd $ findSuggestedHoleFits _message
+{-
+  Relevant bindings include
+    ...
+    extractFitNames :: [Text] -> [Text]
+      (bound at /Users/pepeiborra/scratch/ghcide/src/Development/IDE/Plugin/CodeAction.hs:454:7)
+    (Some bindings suppressed; use -fmax-relevant-binds=N or -fno-max-relevant-binds)
+  Valid hole fits include
+    ...
+    tail :: forall a. [a] -> [a]
+      with tail @Text
+      (imported from ‘Data.List.Extra’ at /Users/pepeiborra/scratch/ghcide/src/Development/IDE/Plugin/CodeAction.hs:45:1-22
+       (and originally defined in ‘GHC.List’))
+    (Some hole fits suppressed; use -fmax-valid-hole-fits=N or -fno-max-valid-hole-fits)
+  Valid refinement hole fits include
+    (++) (_ :: [Text])
+      where (++) :: forall a. [a] -> [a] -> [a]
+      with (++) @Text
+      (imported from ‘Data.List.Extra’ at /Users/pepeiborra/scratch/ghcide/src/Development/IDE/Plugin/CodeAction.hs:45:1-22
+       (and originally defined in ‘GHC.Base’))
+    ...
+    (Some refinement hole fits suppressed; use -fmax-refinement-hole-fits=N or -fno-max-refinement-hole-fits)
+-}
+    | Just [_, holeFits, refinementFits]
+    <- matchRegexSingleLine
+        _message
+        ".*Valid (hole fits|substitutions) include(.*)Valid refinement hole fits include(.*)"
+    , Just holeName <- extractHoleName _message
+    = map (proposeHoleFit holeName)
+    $ fromMaybe [] (extractHoleFits holeFits)
+    ++ fromMaybe [] (extractRefFits refinementFits)
+    | Just [_, holeFits]
+    <- matchRegexSingleLine
+        _message
+        ".*Valid (hole fits|substitutions) include(.*)"
+    , Just fits <- extractHoleFits holeFits
+    , Just holeName <- extractHoleName _message
+    = map (proposeHoleFit holeName) fits
 
     | otherwise = []
+    where
+      extractHoleName = fmap head . flip matchRegex "Found hole: ([^ ]*)"
+      extractHoleFits = fmap (map (!! 1)) . flip matchRegexAll "^    ([^ ]*) :: (.*)$"
+      extractRefFits x = fmap (map (!! 1)) $ matchRegexAll x "^    ([^ ]+( \\(_ :: .+\\))*)$"
+      proposeHoleFit holeName name =
+          ("replace `" <> holeName <> "` with " <> name, [TextEdit _range name])
 
 suggestExtendImport :: Maybe DynFlags -> Maybe T.Text -> Diagnostic -> [(T.Text, [TextEdit])]
 suggestExtendImport (Just dflags) contents Diagnostic{_range=_range,..}
     | Just [binding, mod, srcspan] <-
-      matchRegex _message
+      matchRegexUnifySpaces _message
       "Perhaps you want to add ‘([^’]*)’ to the import list in the import of ‘([^’]*)’ *\\((.*)\\).$"
     , Just c <- contents
     , POk _ (L _ name) <- runParser dflags (T.unpack binding) parseIdentifier
@@ -576,7 +588,7 @@ suggestFixConstructorImport _ Diagnostic{_range=_range,..}
     -- or
     -- import Data.Aeson.Types( Result(..) ) (lsp-ui)
   | Just [constructor, typ] <-
-    matchRegex _message
+    matchRegexUnifySpaces _message
     "‘([^’]*)’ is a data constructor of ‘([^’]*)’ To import it use"
   = let fixedImport = typ <> "(" <> constructor <> ")"
     in [("Fix import of " <> fixedImport, [TextEdit _range fixedImport])]
@@ -620,7 +632,7 @@ suggestConstraint mContents diag@Diagnostic {..}
       findMissingConstraint :: T.Text -> Maybe T.Text
       findMissingConstraint t =
         let regex = "(No instance for|Could not deduce) \\((.+)\\) arising from a use of"
-         in matchRegex t regex <&> last
+         in matchRegexUnifySpaces t regex <&> last
 
 normalizeConstraints :: T.Text -> T.Text -> T.Text
 normalizeConstraints existingConstraints constraint =
@@ -638,7 +650,7 @@ suggestInstanceConstraint contents Diagnostic {..} missingConstraint
 -- • In the expression: x == y
 --   In an equation for ‘==’: (Wrap x) == (Wrap y) = x == y
 --   In the instance declaration for ‘Eq (Wrap a)’
-  | Just [instanceDeclaration] <- matchRegex _message "In the instance declaration for ‘([^`]*)’"
+  | Just [instanceDeclaration] <- matchRegexUnifySpaces _message "In the instance declaration for ‘([^`]*)’"
   = let instanceLine = contents
           & T.splitOn ("instance " <> instanceDeclaration)
           & head & T.lines & length
@@ -657,7 +669,7 @@ suggestInstanceConstraint contents Diagnostic {..} missingConstraint
 --   In an equation for ‘==’:
 --       (Pair x x') == (Pair y y') = x == y && x' == y'
   | Just [instanceLineStr, constraintFirstCharStr]
-    <- matchRegex _message "bound by the instance declaration at .+:([0-9]+):([0-9]+)"
+    <- matchRegexUnifySpaces _message "bound by the instance declaration at .+:([0-9]+):([0-9]+)"
   = let existingConstraints = findExistingConstraints _message
         newConstraints = normalizeConstraints existingConstraints missingConstraint
         instanceLine = readPositionNumber instanceLineStr
@@ -681,7 +693,7 @@ suggestInstanceConstraint contents Diagnostic {..} missingConstraint
         <> "` to the context of the instance declaration"
 
 findTypeSignatureName :: T.Text -> Maybe T.Text
-findTypeSignatureName t = matchRegex t "([^ ]+) :: " <&> head
+findTypeSignatureName t = matchRegexUnifySpaces t "([^ ]+) :: " <&> head
 
 findTypeSignatureLine :: T.Text -> T.Text -> Int
 findTypeSignatureLine contents typeSignatureName =
@@ -727,7 +739,7 @@ suggestFunctionConstraint contents Diagnostic{..} missingConstraint
       findExistingConstraints :: T.Text -> Maybe T.Text
       findExistingConstraints message =
         if message =~ ("from the context:" :: String)
-           then fmap (T.strip . head) $ matchRegex message "\\. ([^=]+)"
+           then fmap (T.strip . head) $ matchRegexUnifySpaces message "\\. ([^=]+)"
            else Nothing
 
       buildNewConstraints :: T.Text -> Maybe T.Text -> T.Text
@@ -782,7 +794,7 @@ removeRedundantConstraints mContents Diagnostic{..}
         & T.lines
         & head
         & T.strip
-        & (`matchRegex` "Redundant constraints?: (.+)")
+        & (`matchRegexUnifySpaces` "Redundant constraints?: (.+)")
         <&> (head >>> parseConstraints)
 
       -- If the type signature is not formatted as expected (arbitrary number of spaces,
@@ -831,7 +843,7 @@ suggestNewImport packageExportsMap ParsedModule {pm_parsed_source = L _ HsModule
           RealSrcLoc s -> Just $ srcLocLine s
           _ -> Nothing
   , insertPos <- Position insertLine 0
-  , extendImportSuggestions <- matchRegex msg
+  , extendImportSuggestions <- matchRegexUnifySpaces msg
     "Perhaps you want to add ‘[^’]*’ to the import list in the import of ‘([^’]*)’"
   = [(imp, [TextEdit (Range insertPos insertPos) (imp <> "\n")])
     | imp <- constructNewImportSuggestions packageExportsMap name extendImportSuggestions
@@ -880,30 +892,23 @@ notInScope (NotInScopeThing t) = t
 
 extractNotInScopeName :: T.Text -> Maybe NotInScope
 extractNotInScopeName x
-  | Just [name] <- matchRegex x "Data constructor not in scope: ([^ ]+)"
+  | Just [name] <- matchRegexUnifySpaces x "Data constructor not in scope: ([^ ]+)"
   = Just $ NotInScopeDataConstructor name
-  | Just [name] <- matchRegex x "Not in scope: data constructor [^‘]*‘([^’]*)’"
+  | Just [name] <- matchRegexUnifySpaces x "Not in scope: data constructor [^‘]*‘([^’]*)’"
   = Just $ NotInScopeDataConstructor name
-  | Just [name] <- matchRegex x "ot in scope: type constructor or class [^‘]*‘([^’]*)’"
+  | Just [name] <- matchRegexUnifySpaces x "ot in scope: type constructor or class [^‘]*‘([^’]*)’"
   = Just $ NotInScopeTypeConstructorOrClass name
-  | Just [name] <- matchRegex x "ot in scope: \\(([^‘ ]+)\\)"
+  | Just [name] <- matchRegexUnifySpaces x "ot in scope: \\(([^‘ ]+)\\)"
   = Just $ NotInScopeThing name
-  | Just [name] <- matchRegex x "ot in scope: ([^‘ ]+)"
+  | Just [name] <- matchRegexUnifySpaces x "ot in scope: ([^‘ ]+)"
   = Just $ NotInScopeThing name
-  | Just [name] <- matchRegex x "ot in scope:[^‘]*‘([^’]*)’"
+  | Just [name] <- matchRegexUnifySpaces x "ot in scope:[^‘]*‘([^’]*)’"
   = Just $ NotInScopeThing name
   | otherwise
   = Nothing
 
 -------------------------------------------------------------------------------------------------
 
-topOfHoleFitsMarker :: T.Text
-topOfHoleFitsMarker =
-#if MIN_GHC_API_VERSION(8,6,0)
-  "Valid hole fits include"
-#else
-  "Valid substitutions include"
-#endif
 
 mkRenameEdit :: Maybe T.Text -> Range -> T.Text -> TextEdit
 mkRenameEdit contents range name =
@@ -1013,11 +1018,31 @@ addBindingToImportList binding importLine = case T.breakOn "(" importLine of
         $  "importLine does not have the expected structure: "
         <> T.unpack importLine
 
+-- | 'matchRegex' combined with 'unifySpaces'
+matchRegexUnifySpaces :: T.Text -> T.Text -> Maybe [T.Text]
+matchRegexUnifySpaces message = matchRegex (unifySpaces message)
+
 -- | Returns Just (the submatches) for the first capture, or Nothing.
 matchRegex :: T.Text -> T.Text -> Maybe [T.Text]
-matchRegex message regex = case unifySpaces message =~~ regex of
+matchRegex message regex = case message =~~ regex of
     Just (_ :: T.Text, _ :: T.Text, _ :: T.Text, bindings) -> Just bindings
     Nothing -> Nothing
+
+-- | Returns all the matchgroups
+matchRegexAll :: T.Text -> T.Text -> Maybe [[T.Text]]
+matchRegexAll message regex = message =~~ regex
+
+matchSingleLine
+    :: (RegexContext Regex T.Text target) => T.Text -> T.Text -> Maybe target
+matchSingleLine regex =
+    matchM (makeRegexOpts blankCompOpt{multiline = False} blankExecOpt regex :: Regex)
+
+-- | Like 'matchRegex' but treat the input as a single line
+--   (dot matches newline, anchors do not treat newline specially)
+matchRegexSingleLine :: T.Text -> T.Text -> Maybe [T.Text]
+matchRegexSingleLine msg regex = do
+  (_ :: T.Text, _ :: T.Text, _ :: T.Text, bindings) <- matchSingleLine regex msg
+  return bindings
 
 setHandlersCodeLens :: PartialHandlers c
 setHandlersCodeLens = PartialHandlers $ \WithMessage{..} x -> return x{

--- a/src/Development/IDE/Plugin/CodeAction.hs
+++ b/src/Development/IDE/Plugin/CodeAction.hs
@@ -559,7 +559,7 @@ processHoleSuggestions mm = (holeSuggestions, refSuggestions)
         mapHead
             (mrAfter . (=~ t " *Valid (hole fits|substitutions) include"))
             validHolesSection
-      let holeFit = T.strip holeFitLine
+      let holeFit = T.strip $ T.takeWhile (/= ':') holeFitLine
       guard (not $ T.null holeFit)
       return holeFit
     refSuggestions = do -- @[]

--- a/src/Development/IDE/Types/Options.hs
+++ b/src/Development/IDE/Types/Options.hs
@@ -94,6 +94,9 @@ data IdeOptions = IdeOptions
     --   Otherwise, return the result of parsing without Opt_Haddock, so
     --   that the parsed module contains the result of Opt_KeepRawTokenStream,
     --   which might be necessary for hlint.
+  , optCustomDynFlags :: DynFlags -> DynFlags
+    -- ^ If given, it will be called right after setting up a new cradle,
+    --   allowing to customize the Ghc options used
   }
 
 data OptHaddockParse = HaddockParse | NoHaddockParse
@@ -157,6 +160,7 @@ defaultIdeOptions session = IdeOptions
     ,optCheckProject = checkProject defaultLspConfig
     ,optCheckParents = checkParents defaultLspConfig
     ,optHaddockParse = HaddockParse
+    ,optCustomDynFlags = id
     }
 
 

--- a/test/exe/Main.hs
+++ b/test/exe/Main.hs
@@ -1508,6 +1508,9 @@ fillTypedHoleTests = let
       , "bar :: Int -> Int -> String"
       , "bar n parameterInt = " <> a <> " (n + " <> b <> " + " <> c <> ")  where"
       , "  localConvert = (flip replicate) 'x'"
+      , ""
+      , "foo :: () -> Int -> String"
+      , "foo = undefined"
 
     ]
 
@@ -1526,32 +1529,35 @@ fillTypedHoleTests = let
     liftIO $ expectedCode @=? modifiedCode
   in
   testGroup "fill typed holes"
-  [ check "replace hole `_` with show"
+  [ check "replace `_` with show"
           "_"    "n" "n"
           "show" "n" "n"
 
-  , check "replace hole `_` with globalConvert"
+  , check "replace `_` with globalConvert"
           "_"             "n" "n"
           "globalConvert" "n" "n"
 
 #if MIN_GHC_API_VERSION(8,6,0)
-  , check "replace hole `_convertme` with localConvert"
+  , check "replace `_convertme` with localConvert"
           "_convertme"   "n" "n"
           "localConvert" "n" "n"
 #endif
 
-  , check "replace hole `_b` with globalInt"
+  , check "replace `_b` with globalInt"
           "_a" "_b"        "_c"
           "_a" "globalInt" "_c"
 
-  , check "replace hole `_c` with globalInt"
+  , check "replace `_c` with globalInt"
           "_a" "_b"        "_c"
           "_a" "_b" "globalInt"
 
 #if MIN_GHC_API_VERSION(8,6,0)
-  , check "replace hole `_c` with parameterInt"
+  , check "replace `_c` with parameterInt"
           "_a" "_b" "_c"
           "_a" "_b"  "parameterInt"
+  , check "replace `_` with foo (_ :: ())"
+          "_" "n" "n"
+          "foo (_ :: ())" "n" "n"
 #endif
   ]
 

--- a/test/exe/Main.hs
+++ b/test/exe/Main.hs
@@ -1529,35 +1529,35 @@ fillTypedHoleTests = let
     liftIO $ expectedCode @=? modifiedCode
   in
   testGroup "fill typed holes"
-  [ check "replace `_` with show"
+  [ check "replace _ with show"
           "_"    "n" "n"
           "show" "n" "n"
 
-  , check "replace `_` with globalConvert"
+  , check "replace _ with globalConvert"
           "_"             "n" "n"
           "globalConvert" "n" "n"
 
 #if MIN_GHC_API_VERSION(8,6,0)
-  , check "replace `_convertme` with localConvert"
+  , check "replace _convertme with localConvert"
           "_convertme"   "n" "n"
           "localConvert" "n" "n"
 #endif
 
-  , check "replace `_b` with globalInt"
+  , check "replace _b with globalInt"
           "_a" "_b"        "_c"
           "_a" "globalInt" "_c"
 
-  , check "replace `_c` with globalInt"
+  , check "replace _c with globalInt"
           "_a" "_b"        "_c"
           "_a" "_b" "globalInt"
 
 #if MIN_GHC_API_VERSION(8,6,0)
-  , check "replace `_c` with parameterInt"
+  , check "replace _c with parameterInt"
           "_a" "_b" "_c"
           "_a" "_b"  "parameterInt"
-  , check "replace `_` with foo (_ :: ())"
+  , check "replace _ with foo _"
           "_" "n" "n"
-          "foo (_ :: ())" "n" "n"
+          "(foo _)" "n" "n"
 #endif
   ]
 

